scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -119,6 +113,53 @@ object TypedPipe extends Serializable {
       override def hash(x: Int): Int = x
     }
   }
+
+   // TODO: CascadingPipe is still tightly bound to cascading
+   case class CascadingPipe[T](pipe: Pipe,
+    fields: Fields,
+    @transient localFlowDef: FlowDef, // not serializable.
+    @transient mode: Mode,
+    converter: TupleConverter[T]) extends TypedPipe[T]
+
+   case class CrossPipe[T, U](left: TypedPipe[T], right: TypedPipe[U]) extends TypedPipe[(T, U)] {
+     def viaHashJoin: TypedPipe[(T, U)] =
```

Comment:
you mean here:
https://github.com/twitter/scalding/pull/1666/files/a76865d69948a2fef0953b8a061a95da9c070bf0#diff-fbe95168deecb5a4b5e9ace30182593eL981

I'm happy to add more optimization rules here, but I think all of those (other than the Merged change) are small potatoes. doing a cross is super rare anyway (because it multiplies the number of outputs) and the optimization for using iterables via closure capture is was not motivated by benchmarks.

The real reason I kept it simple is before we launch this we will want to actually run a lot of more principled and tested optimization rules rather than ad-hoc ones and the ones on cross seemed both ad-hoc (there was even a TODO to remove them) and also rarely exercised (since cross is so rare).